File: src/mind_os_builder/books/validate.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import date
from pathlib import Path
from typing import Any

import yaml


BOOK_STATUSES = frozenset({"reading", "done", "shelved"})
REQUIRED_PROPERTIES = frozenset(
    {"title", "author", "status", "domain", "sources", "created", "updated", "tags"}
)
DATE_PROPERTIES = frozenset({"started", "finished", "created", "updated"})
REQUIRED_BASE_FILTERS = frozenset(
    {
        'file.folder == "wiki/books"',
        'file.ext == "md"',
        'file.name != "density-tracker"',
        '!file.name.startsWith(".")',
        '!file.name.endsWith(".runtime")',
    }
)


@dataclass(frozen=True, slots=True)
class BookIssue:
    code: str
    path: str
    message: str


def _frontmatter(path: Path) -> tuple[dict[str, Any] | None, list[BookIssue]]:
    text = path.read_text(encoding="utf-8")
    if not text.startswith("---\n"):
        return None, [BookIssue("missing_frontmatter", str(path), "缺少 YAML frontmatter")]
    try:
        _, block, _ = text.split("---", 2)
        data = yaml.safe_load(block)
    except (ValueError, yaml.YAMLError) as exc:
        return None, [BookIssue("invalid_frontmatter", str(path), str(exc))]
    if not isinstance(data, dict):
        return None, [BookIssue("invalid_frontmatter", str(path), "frontmatter 必须是对象")]
    return data, []


def _is_iso_date(value: object) -> bool:
    if isinstance(value, date):
        return True
    if not isinstance(value, str):
        return False
    try:
        return date.fromisoformat(value).isoformat() == value
    except ValueError:
        return False


def validate_book_page(path: Path) -> list[BookIssue]:
    data, issues = _frontmatter(path)
    if data is None:
        return issues

    for key in sorted(REQUIRED_PROPERTIES):
        if key not in data or data[key] is None or data[key] == "":
            issues.append(BookIssue("missing_property", str(path), f"缺少属性：{key}"))

    status = data.get("status")
    if status is not None and status not in BOOK_STATUSES:
        issues.append(BookIssue("invalid_status", str(path), f"未知阅读状态：{status}"))

    for key in sorted(DATE_PROPERTIES):
        value = data.get(key)
        if value not in (None, "") and not _is_iso_date(value):
            issues.append(BookIssue("invalid_date", str(path), f"{key} 必须是 YYYY-MM-DD"))

    sources = data.get("sources")
    if sources is not None and (not isinstance(sources, int) or isinstance(sources, bool) or sources < 0):
        issues.append(BookIssue("invalid_sources", str(path), "sources 必须是非负整数"))

    tags = data.get("tags")
    if tags is not None and (not isinstance(tags, list) or not all(isinstance(tag, str) for tag in tags)):
        issues.append(BookIssue("invalid_tags", str(path), "tags 必须是字符串列表"))
    return issues


def validate_base_definition(path: Path) -> list[BookIssue]:
    try:
        data = yaml.safe_load(path.read_text(encoding="utf-8"))
    except (OSError, yaml.YAMLError) as exc:
        return [BookIssue("invalid_base", str(path), str(exc))]
    if not isinstance(data, dict):
        return [BookIssue("invalid_base", str(path), ".base 必须是 YAML 对象")]

    filters = data.get("filters")
    expressions = filters.get("and") if isinstance(filters, dict) else None
    if not isinstance(expressions, list) or not all(isinstance(item, str) for item in expressions):
        return [BookIssue("unsafe_base_filter", str(path), "缺少顶层 and 过滤")]
    if set(expressions) != REQUIRED_BASE_FILTERS:
        return [
            BookIssue(
                "unsafe_base_filter",
                str(path),
                "Base 必须严格限定 wiki/books 与 Markdown，并排除运行态文件",
            )
        ]
    return []


def _is_runtime_page(path: Path) -> bool:
    return (
        path.name.startswith(".")
        or path.stem == "density-tracker"
        or path.stem.endswith(".runtime")
    )


def validate_books_module(vault_root: Path) -> list[BookIssue]:
    books = vault_root / "wiki/books"
    base = books / "books.base"
    issues: list[BookIssue] = []
    if not base.is_file():
        issues.append(BookIssue("missing_base", str(base), "缺少 books.base"))
    else:
        issues.extend(validate_base_definition(base))

    if not books.is_dir():
        issues.append(BookIssue("missing_books_directory", str(books), "缺少 wiki/books"))
        return issues
    for page in sorted(books.glob("*.md")):
        if _is_runtime_page(page):
            issues.append(
                BookIssue("runtime_file_in_books", str(page), "运行态 Markdown 不得进入 Book Base 目录")
            )
            continue
        issues.extend(validate_book_page(page))
    return issues

File: src/books/validate.ts
import { readdir, readFile } from "node:fs/promises";
import { basename, join } from "node:path";
import { parse } from "yaml";
import { parseFrontmatter } from "../lib/frontmatter.js";
import { resolveReadPath } from "../lib/paths.js";
import { blockedFromError, noopResult, previewResult, type CliResult } from "../lib/result.js";

const statuses = new Set(["reading", "done", "shelved"]);
const requiredProperties = ["title", "author", "status", "domain", "sources", "created", "updated", "tags"];
const baseFilters = new Set([
  'file.folder == "wiki/books"',
  'file.ext == "md"',
  'file.name != "density-tracker"',
  '!file.name.startsWith(".")',
  '!file.name.endsWith(".runtime")',
]);

type Issue = { code: string; path: string; message: string };

function isIsoDate(value: unknown): boolean {
  if (typeof value !== "string") {
    return false;
  }
  const match = /^(\d{4})-(\d{2})-(\d{2})$/u.exec(value);
  if (match === null) {
    return false;
  }
  const year = Number(match[1]);
  const month = Number(match[2]);
  const day = Number(match[3]);
  const date = new Date(Date.UTC(year, month - 1, day));
  return date.getUTCFullYear() === year && date.getUTCMonth() === month - 1 && date.getUTCDate() === day;
}

function pageIssues(relative: string, content: string): Issue[] {
  const issues: Issue[] = [];
  const parsed = parseFrontmatter(content);
  if (!parsed.ok && parsed.reason === "missing") {
    return [{ code: "missing_frontmatter", path: relative, message: "缺少 YAML frontmatter" }];
  }
  if (!parsed.ok && parsed.reason === "unclosed") {
    return [{ code: "invalid_frontmatter", path: relative, message: "frontmatter 未闭合" }];
  }
  if (!parsed.ok) {
    return [{ code: "invalid_frontmatter", path: relative, message: "frontmatter 无效" }];
  }
  const metadata = parsed.metadata;
  if (typeof metadata !== "object" || metadata === null) {
    return [{ code: "invalid_frontmatter", path: relative, message: "frontmatter 必须是对象" }];
  }
  const page = metadata as Record<string, unknown>;
  for (const property of requiredProperties) {
    if (page[property] === undefined || page[property] === null || page[property] === "") {
      issues.push({ code: "missing_property", path: relative, message: `缺少属性：${property}` });
    }
  }
  if (page.status !== undefined && (typeof page.status !== "string" || !statuses.has(page.status))) {
    issues.push({ code: "invalid_status", path: relative, message: "未知阅读状态" });
  }
  for (const property of ["started", "finished", "created", "updated"]) {
    if (page[property] !== undefined && page[property] !== null && page[property] !== "" && !isIsoDate(page[property])) {
      issues.push({ code: "invalid_date", path: relative, message: `${property} 必须是 YYYY-MM-DD` });
    }
  }
  if (page.sources !== undefined && (!Number.isInteger(page.sources) || typeof page.sources !== "number" || page.sources < 0)) {
    issues.push({ code: "invalid_sources", path: relative, message: "sources 必须是非负整数" });
  }
  if (page.tags !== undefined && (!Array.isArray(page.tags) || !page.tags.every((tag) => typeof tag === "string"))) {
    issues.push({ code: "invalid_tags", path: relative, message: "tags 必须是字符串列表" });
  }
  return issues;
}

function baseIssues(relative: string, content: string): Issue[] {
  try {
    const data: unknown = parse(content);
    const expressions = typeof data === "object" && data !== null && typeof (data as { filters?: unknown }).filters === "object" && (data as { filters: { and?: unknown } }).filters !== null
      ? (data as { filters: { and?: unknown } }).filters.and
      : undefined;
    if (!Array.isArray(expressions) || !expressions.every((item) => typeof item === "string") || expressions.length !== baseFilters.size || !expressions.every((item) => baseFilters.has(item))) {
      return [{ code: "unsafe_base_filter", path: relative, message: "Base 必须严格限定 wiki/books 与 Markdown" }];
    }
    return [];
  } catch {
    return [{ code: "invalid_base", path: relative, message: "books.base 不是有效 YAML" }];
  }
}

export async function validateBooks(root: string): Promise<CliResult> {
  try {
    const books = await resolveReadPath(root, "wiki/books");
    const entries = await readdir(books, { withFileTypes: true });
    const issues: Issue[] = [];
    const base = entries.find((entry) => entry.name === "books.base");
    if (base?.isFile()) {
      issues.push(...baseIssues("wiki/books/books.base", await readFile(join(books, "books.base"), "utf8")));
    } else {
      issues.push({ code: "missing_base", path: "wiki/books/books.base", message: "缺少 books.base" });
    }
    for (const entry of entries) {
      if (!entry.isFile() || !entry.name.endsWith(".md")) {
        continue;
      }
      const relative = `wiki/books/${entry.name}`;
      if (entry.name.startsWith(".") || basename(entry.name, ".md") === "density-tracker" || basename(entry.name, ".md").endsWith(".runtime")) {
        issues.push({ code: "runtime_file_in_books", path: relative, message: "运行态 Markdown 不得进入 Book Base 目录" });
      } else {
        issues.push(...pageIssues(relative, await readFile(join(books, entry.name), "utf8")));
      }
    }
    const data = { issue_count: issues.length, issues };
    return issues.length === 0 ? noopResult(data) : previewResult(data);
  } catch (error: unknown) {
    if ((error as NodeJS.ErrnoException).code === "ENOENT") {
      return previewResult({ issue_count: 2, issues: [
        { code: "missing_base", path: "wiki/books/books.base", message: "缺少 books.base" },
        { code: "missing_books_directory", path: "wiki/books", message: "缺少 wiki/books" },
      ] });
    }
    return blockedFromError(error, {});
  }
}
